src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/BatchingWorkQueue.cs

```diff
@@ -0,0 +1,263 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.CodeAnalysis.Razor.Workspaces
+{
+    internal sealed class BatchingWorkQueue : IDisposable
+    {
+        private readonly Dictionary<string, BatchableWorkItem> _work;
+        private readonly TimeSpan _batchingTimeSpan;
+        private readonly ErrorReporter _errorReporter;
+        private readonly CancellationTokenSource _disposalCts;
+        private Timer? _timer;
+        private bool _disposed;
+
+        public BatchingWorkQueue(
+            TimeSpan batchingTimeSpan,
+            StringComparer keyComparer,
+            ErrorReporter errorReporter)
+        {
+            if (keyComparer is null)
+            {
+                throw new ArgumentNullException(nameof(keyComparer));
+            }
+
+            if (errorReporter is null)
+            {
+                throw new ArgumentNullException(nameof(errorReporter));
+            }
+
+            _batchingTimeSpan = batchingTimeSpan;
+            _errorReporter = errorReporter;
+            _disposalCts = new CancellationTokenSource();
+            _work = new Dictionary<string, BatchableWorkItem>(keyComparer);
+        }
+
+        private bool IsScheduledOrRunning => _timer != null;
+
+        // Used in unit tests to ensure we can control when background work starts.
+        private ManualResetEventSlim? BlockBackgroundWorkStart { get; set; }
+
+        // Used in unit tests to ensure we can know when background work finishes.
+        private ManualResetEventSlim? NotifyBackgroundWorkStarting { get; set; }
+
+        // Used in unit tests to ensure we can know when background has captured its current workload.
+        private ManualResetEventSlim? NotifyBackgroundCapturedWorkload { get; set; }
+
+        // Used in unit tests to ensure we can control when background work completes.
+        private ManualResetEventSlim? BlockBackgroundWorkCompleting { get; set; }
+
+        // Used in unit tests to ensure we can know when background work finishes.
+        private ManualResetEventSlim? NotifyBackgroundWorkCompleted { get; set; }
+
+        // Used in unit tests to ensure we can know when errors are reported
+        private ManualResetEventSlim? NotifyErrorBeingReported { get; set; }
+
+        public void Enqueue(string key, BatchableWorkItem workItem)
+        {
+            lock (_work)
+            {
+                if (_disposed)
+                {
+                    return;
```

Comment:
Hmmm, could but it doesn't buy us much. Reasons I can come up with for it being valuable include ensuring callers handle their own dispose correctly; however, I don't see any negative drawback to calling this and no-oping.

Reply:
Yeah, I asked about this because it want it to be obvious when something is trying to distribute work to a "dead" object in case it somehow happens outside the "VS is already shutting down" scenario. I don't feel SUPER strongly that we gotta do this, I just prefer to fail fast and loud.

Reply:
> I don't feel SUPER strongly that we gotta do this, I just prefer to fail fast and loud.

Will give it a shot and see where it leads 😄